saleor/checkout/utils.py

```diff
@@ -175,18 +184,29 @@ def add_variants_to_checkout(
     country_code = checkout.get_country()
 
     checkout_lines = checkout.lines.select_related("variant")
-    variant_ids_in_lines = {line.variant_id: line for line in checkout_lines}
+
+    lines_by_variant_id = {line.variant_id: line for line in checkout_lines}
+    lines_by_ids = {str(line.pk): line for line in checkout_lines}
+    variants_map = {variant.pk: variant for variant in variants}
+
     to_create: List[CheckoutLine] = []
     to_update: List[CheckoutLine] = []
     to_delete: List[CheckoutLine] = []
-    for variant, line_data in zip(variants, checkout_lines_data):
+
+    for line_data in checkout_lines_data:
         _append_line_to_update(
-            to_update, to_delete, variant, line_data, replace, variant_ids_in_lines
+            to_update, to_delete, line_data, replace, lines_by_variant_id, lines_by_ids
```

Comment:
Also, do we really need to pass two separate dicts that store the line details? 
Maybe I am missing something but can't we unify the functions that operate on lines to use only line IDs instead of trying to use lineID and then trying to use variantID?

Reply:
Right, I did that to handle cases where line has a variant that is unavailable - `fetch_checkout_lines` was not returning `CheckoutLineInfo` for those lines so I was not able to find line for provided variant. 
But now I added flag `skip_lines_with_unavailable_variants`  to `fetch_checkout_lines` function and removed `lines_by_variant_id`  to operate only on `line_id` as you suggested.